File: 198.打家劫舍.cpp
/*
 * @lc app=leetcode.cn id=198 lang=cpp
 *
 * [198] 打家劫舍
 */

// @lc code=start

#include <vector>
#include <algorithm>

class Solution {
public:
    int rob(std::vector<int>& nums) {
        if(nums.empty()) return 0;
        std::pair<int, int> last = {0, nums[0]};

        for(int i = 1; i < nums.size(); ++i){
            last = {
                std::max(last.first, last.second),
                last.first + nums[i]
            };
        }

        return std::max(last.first, last.second);
    }
};
// @lc code=end

File: 198.打家劫舍.rs
/*
 * @lc app=leetcode.cn id=198 lang=rust
 *
 * [198] 打家劫舍
 */

// @lc code=start
impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        let mut dp = vec![[0;2];nums.len()];
        dp[0][0] = 0;
        dp[0][1] = nums[0];
        for idx in 1..nums.len() {
            dp[idx][0] = dp[idx - 1][0].max(dp[idx - 1][1]);
            dp[idx][1] = dp[idx - 1][0] + nums[idx];
        }
        return dp[dp.len() - 1][0].max(dp[dp.len() - 1][1]);
    }
}
// @lc code=end

